sycl/unittests/helpers/PiMock.hpp

```diff
@@ -0,0 +1,197 @@
+    // Save a copy of the platform resource
+    MPlatform = OriginalPlatform;
+  }
+
+  /// Constructs PiMock from a device_selector, provided that
+  /// a non-host device can and will be selected.
+  ///
+  /// \param DevSelector is a reference to a device_selector instance.
+  explicit PiMock(const cl::sycl::device_selector &DevSelector)
+      : MPlatform{DevSelector} {
+    assert(!MPlatform.is_host() && "PI mock isn't supported for host");
+    auto ImplPtr = detail::getSyclObjImpl(MPlatform);
+    const RT::PiPlugin &PiPluginRef = ImplPtr->getPlugin().getPiPlugin();
+    // Extract the PiPlugin instance by a non-const pointer,
+    // explicitly allowing modification
+    PiPluginMockPtr = &const_cast<RT::PiPlugin &>(PiPluginRef);
+  }
+
+  /// Explicit construction from a host_selector is forbidden.
+  PiMock(const cl::sycl::host_selector &HostSelector) = delete;
+
+  PiMock(const PiMock &) = delete;
+  PiMock &operator=(const PiMock &) = delete;
+  ~PiMock() = default;
+
+  /// Returns a handle to the SYCL platform instance.
+  ///
+  /// \return A reference to the SYCL platform.
+  cl::sycl::platform &getPlatform() { return MPlatform; }
+
+  template <detail::PiApiKind PiApiOffset>
+  using FuncPtrT = typename RT::PiFuncInfo<PiApiOffset>::FuncPtrT;
+  template <detail::PiApiKind PiApiOffset>
+  using SignatureT = typename std::remove_pointer<FuncPtrT<PiApiOffset>>::type;
+
+  /// Redefines the implementation of a given PI API to the input
+  /// function object.
+  ///
+  /// \param Replacement is a mock std::function instance to be
+  ///        called instead of the given PI API. It is imperative
+  ///        that this function be not constructed from a lambda.
+  template <detail::PiApiKind PiApiOffset>
+  void redefine(const std::function<SignatureT<PiApiOffset>> &Replacement) {
+    // TODO: Find a way to store FPointer first so that real PI functions can
+    // be called alongside the mock ones. Something like:
+    // `enum class MockPIPolicy { InsteadOf, Before, After};`
+    // may need to be introduced.
+    auto *FuncPtr = Replacement.template target<FuncPtrT<PiApiOffset>>();
+    assert(FuncPtr &&
+           "Function target is empty, try passing a lambda directly");
+    setFuncPtr<PiApiOffset>(PiPluginMockPtr, *FuncPtr);
+  }
+
+  /// A `redefine` overload for function pointer/captureless lambda
+  /// arguments.
+  ///
+  /// \param Replacement is a mock callable assignable to a function
+  ///        pointer (function pointer/captureless lambda).
+  template <detail::PiApiKind PiApiOffset, typename FunctorT>
+  void redefine(const FunctorT &Replacement) {
+    // TODO: Check for matching signatures/assignability
+    setFuncPtr<PiApiOffset>(PiPluginMockPtr, Replacement);
+  }
+
+private:
+  cl::sycl::platform MPlatform;
+  // Extracted at initialization for convenience purposes. The resource
+  // itself is owned by the platform instance.
+  RT::PiPlugin *PiPluginMockPtr;
```

Comment:
Suggest adding M at the beggining of the member name to align with MPlatform

Reply:
Done